component-catalog-connectors/artifactory-connector/README.md

```diff
@@ -0,0 +1,203 @@
+> outputs: []
+> implementation: {}
+> ```
+
+Elyra only stores pointers to the source-catalog of each component in its `.pipeline` files.
+This means your catalogs must continue to include ALL old component versions, otherwise old pipelines will be unable to run.
+
+You can solve this problem with the `Artifactory Catalog Connector` by adding TWO catalog instances for your single Artifactory repo:
+
+1. that has only the latest version of each component:
+    - `max_files_per_folder = 1`
+    - `file_ordering = VERSION_DESCENDING`
+2. that has all versions of each component:
+    - `max_files_per_folder = -1`
+
+> 🟦 __Tip__ 🟦
+>
+> Name your component YAML files with consistent prefixes before the version.
+> 
+> This is because `VERSION_DESCENDING` treats the whole file-name as a version.
+> For example, `aaaa-1.0.0.yaml` would be sorted before `bbbb-9.0.0.yaml`.
+>
+> Alternatively, don't include any prefix on file names, for example `1.0.0.yaml`, `1.1.0.yaml`.
+
+
+### Example Configs
+
+Assume an artifactory server `https://example.com/artifactory/` has a repository called `elyra-components` 
+with the following folder structure:
+
+```
+component_1/
+   __COMPONENT__
+   component-1.0.9.yaml
+   component-1.0.10.yaml
+component_2/
+   hidden_component/
+      __COMPONENT__
+      component-1.0.0.yaml
+      component-1.1.0.yaml
+   __COMPONENT__
+   component-1.0.0.yaml
+   component-1.1.0.yaml
+component_3/
+   component-1.0.0.yaml
+   component-1.1.0.yaml
+```
+
+__Example 1:__
+
+```
+Configs:
+=========
+artifactory_url      = https://example.com/artifactory/
+repository_name      = elyra-components
+repository_path      = /
+max_recursion_depth  = 3
+max_files_per_folder = -1
+file_filter          = *.yaml
+file_ordering        = VERSION_DESCENDING
+
+Matched:
+=========
+https://example.com/artifactory/elyra-components/component_1/component-1.0.9.yaml
+https://example.com/artifactory/elyra-components/component_1/component-1.0.10.yaml
+https://example.com/artifactory/elyra-components/component_2/component-1.0.0.yaml
+https://example.com/artifactory/elyra-components/component_2/component-1.1.0.yaml
+
+Notes:
+=========
+- the `component_3/` files are not matched as this folder does not contain a `__COMPONENT__` marker
+- the `component_2/hidden_component/` files are not matched as recursion stops at the first `__COMPONENT__` marker
+```
+
+__Example 2:__
+
+```
+Configs:
+=========
+artifactory_url      = https://example.com/artifactory/
+repository_name      = elyra-components
+repository_path      = /
+max_recursion_depth  = 0
+max_files_per_folder = -1
+file_filter          = *.yaml
+file_ordering        = VERSION_DESCENDING
+
+Matched:
+=========
+N/A
+
+Notes:
+=========
+- no files are matched, as `max_recursion_depth` is `0`
+```
+
+__Example 3:__
+
+```
+Configs:
+=========
+artifactory_url      = https://example.com/artifactory/
+repository_name      = elyra-components
+repository_path      = /
+max_recursion_depth  = 3
```

Comment:
@ptitzler two questions:

1. what was your folder structure?
2. were you testing using the test framework, or with an actual artifactory?

NOTE: the recursion stops for a particular folder branch once the first `__COMPONENT__` file is encountered, did you accidentally include a `__COMPONENT__` at the root of your repo?

Reply:
You are right, it was a user error. The file was present in the root. After I've removed it everything seemed to work as expected. What scenario where you thinking about when you decided that no further recursive search should be performed once this file was detected?

Reply:
@ptitzler it's a protection mechanism to prevent unnecessary recursion, and as long as people understand that a `__COMPONENT__` marker prevents further recursion (which is highlighted in the README), I think it should be fine?